### PR TITLE
Fix cpp node instance check

### PR DESCRIPTION
**Pull request by albanD**
Mostly visible when calling multi_grad_hook and thus using this to test it.

### CODE REVIEW

**Review by jackiexu1992 — APPROVED** [review]

File: torch/csrc/autograd/python_cpp_function.cpp [hunk 1]
```diff
@@ -255,11 +255,9 @@ struct DefaultFunctionType {
   }
 
   PyTypeObject type;
-};
+} default_type;
```
Comment (ezyang) [inline torch/csrc/autograd/python_cpp_function.cpp:258]: This doesn't look right. You're static initializing default_type now, how do you guarantee you have the GIL?

Comment (albanD) [reply torch/csrc/autograd/python_cpp_function.cpp]: Sam's answer was not a straight yes so updated.

File: torch/csrc/autograd/python_cpp_function.cpp [hunk 2]
```diff
@@ -227,6 +227,7 @@ PyTypeObject* _initFunctionPyTypeObject(
     const char* name,
     PyGetSetDef* function_properties,
     PyMethodDef* function_methods) {
+  type.ob_base = { PyObject_HEAD_INIT(nullptr) 0 };
```
Comment (albanD) [inline torch/csrc/autograd/python_cpp_function.cpp:230]: This is asked by Sam to actually initialize the header properly and make the objects immortal (since they are static) and this will init the refcount to 1 with cpython <3.12.
The incref here is removed but the ones for the objects created via addSubclass<>() are left since they call API that steal references.

**Comment by albanD** [comment]
(forgot to commit everything before pushing)

File: torch/csrc/autograd/python_cpp_function.cpp [hunk 3]
```diff
@@ -251,15 +251,17 @@ static std::unordered_set<PyTypeObject*> cpp_function_types_set;
 struct DefaultFunctionType {
   DefaultFunctionType() : type() {
     _initFunctionPyTypeObject(type, "CppFunction", nullptr, nullptr);
-    Py_INCREF(&type);
```
Comment (ezyang) [inline torch/csrc/autograd/python_cpp_function.cpp:254]: You don't need the incref?

**Review by ezyang — APPROVED** [review]
This looks plausible, IG? Maybe @colesbury would be better able to review?

File: torch/csrc/autograd/python_cpp_function.cpp (hunk 3 above)
Comment (albanD) [reply torch/csrc/autograd/python_cpp_function.cpp:254]: Correct, see my comment above: https://github.com/pytorch/pytorch/pull/125875#discussion_r1596088552
tl;dr: now that we properly initialize the header in _initFunctionPyTypeObject (was kept all 0s before), this function sets the refcount to 1 properly.

**Comment by albanD** [comment]
@pytorchbot merge

I prepared the patch with @colesbury so I'll just merge now.
Do let me know if there is anything unexpected though and I can just revert!
